Complete lab06

n0npax opened the pull request:
Implement CRUD for pet structure
Implement Sync and mem storer
Implement tests

Fixes #356 

Review of colleague's PR #352 

#### Changes proposed in this PR:

## Complete lab06
Implement CRUD for puppy structure
Implement tests

n0npax committed:
Complete lab06

Implement CRUD for pet structure
Implement Sync and mem storer
Implement tests

runnerdave commented on 06_puppy/n0npax/store_test.go:
```diff
@@ -0,0 +1,121 @@
+	}
+}
+
+func (s *storerSuite) TestReadPuppy_IDDoesNotExist() {
```
naming is a bit inconsistent, this test name has an underscore while the others do not

n0npax replied:
You're right, I should avoid underscores in names. Fixed

runnerdave commented on 06_puppy/n0npax/syncStore.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func (m *SyncStore) CreatePuppy(puppy *Puppy) int {
+	// generate new id
```
unnecessary comment

n0npax replied:
fixed. Thanks

runnerdave commented on 06_puppy/n0npax/memStore.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func (m MemStore) DeletePuppy(id int) (bool, error) {
+
```
minor, but this empty line is not needed

n0npax replied:
fixed. Thanks

runnerdave reviewed (dismissed):
good job, very minor suggestions

rokane commented on 06_puppy/n0npax/types.go:
```diff
@@ -0,0 +1,25 @@
+import (
+	"sync"
+)
+
```
Great work separating the types out into their own file. Are you able to add some documentation to these exported types? It will be good to get in the habit of documenting any exported types, interfaces and functions as we start to move towards building an API.

General convention is a comment above the exported type/function, which starts with the type/function name. Eg.

```go
// Puppy contains information about a single Puppy ...
type Puppy struct {
    ...
}
```

n0npax replied:
Fixed. Thanks!

rokane commented on 06_puppy/n0npax/types.go:
```diff
@@ -0,0 +1,25 @@
+}
+
+type Storer interface {
+	ReadPuppy(ID int) (*Puppy, error)
```
Minor little nit. But the naming of input parameters here is inconsistent. We have `ID` and `id` across the different interface functions. Pick one and use that consistently.

n0npax replied:
Fixed. All interfaces have capitalized `ID` now. Thanks!

rokane commented on 06_puppy/n0npax/memStore.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+func (m MemStore) ReadPuppy(id int) (*Puppy, error) {
+	puppy, ok := m[id]
```
If you flip the logic here, you can save a line of code.

```go
if puppy, ok := m[id]; ok {
    return puppy, nil
}
return nil, errors.New("doesn't exist")
```

n0npax replied:
fixed. Thanks

rokane commented on 06_puppy/n0npax/memStore.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+func (m MemStore) UpdatePuppy(id int, puppy *Puppy) error {
+	m[id] = puppy
```
Can you include an error check here to see if the puppy exists? Since we are returning an error, it suggests we should be doing this kind of check.
```go
if _, ok := m[id]; !ok {
    return errors.New(fmt.Sprintf("Puppy with ID: %d does not exist", id))
}
m[id] = puppy
return nil
```

n0npax replied:
fixed. Thanks for that!

rokane commented on 06_puppy/n0npax/syncStore.go:
```diff
@@ -0,0 +1,42 @@
+	return &SyncStore{}
+}
+
+func (m *SyncStore) CreatePuppy(puppy *Puppy) int {
```
Here you should be making use of the `sync.Mutex` capability to put a mutex lock on the resource to ensure race conditions can't occur. If you extend your definition of the `SyncStore` you will be able to do this.

```go
type SyncStore struct {
    sync.Map
    sync.Mutex
}

func (m *SyncStore) CreatePuppy(puppy *Puppy) int {
    m.Lock()
    defer m.Unlock()

    // Continue with functionality below
}

```

n0npax replied:
`create`, `update` & `delete` are using mutex now. I believe Read can be done without a mutex.

rokane commented on 06_puppy/n0npax/syncStore.go:
```diff
@@ -0,0 +1,42 @@
+
+func (m *SyncStore) CreatePuppy(puppy *Puppy) int {
+	var newID int
+	m.Range(func(_, _ interface{}) bool {
```
I also don't think it is required to iterate over the map everytime, we can turn this from a linear, to a constant runtime by storing the number of items as another variable in the `SyncStore`.

```go
type SyncStore struct {
    sync.Map
    sync.Mutex
    total    int
}
```

rokane commented on 06_puppy/n0npax/syncStore.go:
```diff
@@ -0,0 +1,42 @@
+	return &puppy, nil
+}
+
+func (m *SyncStore) UpdatePuppy(id int, puppy *Puppy) error {
```
Again here, I think there should be a check to see if the id exists, and if not a suitable error condition

n0npax replied:
fixed. Thanks!

rokane commented on 06_puppy/n0npax/syncStore.go:
```diff
@@ -0,0 +1,42 @@
+	return puppy.ID
+}
+
+func (m *SyncStore) ReadPuppy(id int) (*Puppy, error) {
```
Again can flip the logic and save a line here as well. Same as described in the memstore

n0npax replied:
fixed. Thanks!

rokane commented on 06_puppy/n0npax/store_test.go:
```diff
@@ -0,0 +1,121 @@
+
+const (
+	syncStorer storerImpl = iota
+	memStorer  storerImpl = iota
```
You don't need to specify this as two `iota`'s. You can just define this as below and they will be seperate enum values.

```go
const (
    syncStorer storerImpl = iota
    memStorer
)
```

n0npax replied:
fixed. Thanks

rokane commented on 06_puppy/n0npax/store_test.go:
```diff
@@ -0,0 +1,121 @@
+)
+
+var (
+	puppy0 = func() Puppy {
```
Nice work 👍

n0npax replied:
:rocket: Thanks!

rokane commented on 06_puppy/n0npax/store_test.go:
```diff
@@ -0,0 +1,121 @@
+
+func (s *storerSuite) SetupTest() {
+	switch s.impl {
+	case syncStorer:
```
Awesome work!! Great use of the SetupTest function.

Can you please extract the inserting of a puppy so it is outside the switch statement, that way you won't have to repeat it for each implementation. Saves a couple of lines of code.

n0npax replied:
fixed Thanks for that!

rokane commented on 06_puppy/n0npax/store_test.go:
```diff
@@ -0,0 +1,121 @@
+	suite.Run(t, &syncSuite)
+}
+
+func TestStorerMemoryMap(t *testing.T) {
```
You can combine these two functions into one, and just run the suite against the different configurations. Also since you are using the `SetupTest` method, you are ensuring in there that the store is set. You should be able to remove this and delegate it to that function as a pre run task.

```go
func TestStorer(t *testing.T) {
    suite.Run(t, &storerSuite{impl: syncStorer})
    suite.Run(t, &storerSuite{impl: memStorer})
}
```

n0npax replied:
Fixed. Good hint. Thanks!

rokane reviewed (changes requested):
Awesome work. Your tests and code are looking really succinct and clean. I have suggested a few tidy up tasks throughout, can you please address them and let me know, I will aim to look at it so you can continue with the other lab.

n0npax committed:
Impelent CR fixes

Improved error messages
Optimalize tests
Improve syntax
Add comments to exported types

n0npax commented:
Added `ID` corruption check in the midtime. I'd like to disallow updating `ID` to prevent the situation when the object is under `id`: 8 but internal `id`: is 9

rokane reviewed (approved):
Looks good, keep it up!!

n0npax committed:
Merge branch 'master' into lab06
